linkerd/app/outbound/src/tls/detect.rs

```diff
@@ -0,0 +1,92 @@
+use linkerd_app_core::{
+    io,
+    svc::{self, layer, ServiceExt},
+    tls::{
+        server::{detect_sni, DetectIo},
+        ServerName,
+    },
+    Error, Result,
+};
+use std::{
+    future::Future,
+    pin::Pin,
+    task::{Context, Poll},
+};
+use thiserror::Error;
+use tokio::time;
+
+#[derive(Clone, Debug, Error)]
+#[error("SNI detection timed out")]
+pub struct SniDetectionTimeoutError;
+
+#[derive(Clone, Debug, Error)]
+#[error("Could not find SNI")]
+pub struct NoSniFoundError;
+
+#[derive(Clone, Debug, Default)]
+pub struct NewDetectSniServer<N> {
+    inner: N,
+    timeout: time::Duration,
+}
+
+#[derive(Clone, Debug, Default)]
+pub struct DetectSniServer<T, N> {
+    target: T,
+    inner: N,
+    timeout: time::Duration,
+}
+
+impl<N> NewDetectSniServer<N> {
+    pub fn layer() -> impl layer::Layer<N, Service = Self> + Copy {
+        let timeout = time::Duration::from_secs(10);
+        layer::mk(move |inner| Self { inner, timeout })
+    }
+}
+
+impl<T, N: Clone> svc::NewService<T> for NewDetectSniServer<N> {
+    type Service = DetectSniServer<T, N>;
+
+    fn new_service(&self, target: T) -> Self::Service {
+        DetectSniServer {
+            target,
+            timeout: self.timeout,
+            inner: self.inner.clone(),
+        }
+    }
+}
+
+impl<T, I, N, S> svc::Service<I> for DetectSniServer<T, N>
+where
+    T: Clone + Send + Sync + 'static,
+    I: io::AsyncRead + io::Peek + io::AsyncWrite + Send + Sync + Unpin + 'static,
+    N: svc::NewService<(ServerName, T), Service = S> + Clone + Send + 'static,
+    S: svc::Service<DetectIo<I>> + Send,
```

Comment:
This DetectIo is probably wrong here -- we never want to proceed if we don't detect an SNI, right? So we can be sure we're handling a PrefixedIo

Reply:
Not sure this is correct. The "eitherness" of `DetectIo` does not distinguish between whether an SNI was detected or not. A `DetectIo` is a PrefixedIO in the cases when we could not parse the SNI via peeking. In these cases, we read from the socket and then construct a PrefixedIo. In the case of peeking a DetectIO is the original IO... 

So I think a `DetectIo` makes sense here(i.e detection has been performed on this IO regardless of the outcome.)

Does that sound right ?